### PR TITLE
Fix epoch block resync on start problem

### PR DESCRIPTION
**Pull request by Ujang360**
Also bump the version up to `v0.19.0.1` and use `GCC 8.3.0` to build it.

## Explaination
The main cause of the `resync problem` is that the previous Dockerfile and its script tried to find non-existence user in the host. This also means that the docker image was not portable.

I can't find a better solution than using the `root` user. Do you guys think this would lead into another problem?

#### Build the image first

```bash
docker build --pull -t build-test/bitcoind:0.19.0.1 .
```

#### Create `/opt/bitcoin/bitcoin.conf` with this configuration

```conf
# This config should be placed in following path:
# /opt/bitcoin/bitcoin.conf

# [core]
# Maintain a full transaction index, used by the getrawtransaction rpc call.
txindex=1

# [network]
# Automatically create Tor hidden service.
listenonion=0
# Only connect to peers via IPV4
onlynet=ipv4

# [zeromq]
# Enable publishing of block hashes to <address>.
zmqpubhashblock=tcp://0.0.0.0:28332
# Enable publishing of transaction hashes to <address>.
zmqpubhashtx=tcp://0.0.0.0:28333
# Enable publishing of raw block hex to <address>.
zmqpubrawblock=tcp://0.0.0.0:28334
# Enable publishing of raw transaction hex to <address>.
zmqpubrawtx=tcp://0.0.0.0:28335

# [debug]
# Enable debug logging for all categories.
#debug=1
# Log IP Addresses in debug output.
logips=1
# Log timestamps with microsecond precision.
logtimemicros=1
# Log transaction fee per kB when mining blocks.
printpriority=1
# Run this node on the Bitcoin Test Network. Equivalent to -chain=test
testnet=1

# [wallet]
# If paytxfee is not set, include enough fee so that transactions should confirm within <n> blocks
txconfirmtarget=3

# [Sections]
# Most options automatically apply to mainnet, testnet, and regtest networks.
# If you want to confine an option to just one network, you should add it in the relevant section.
# EXCEPTIONS: The options addnode, connect, port, bind, rpcport, rpcbind and wallet
# only apply to mainnet unless they appear in the appropriate section below.

# Options only for mainnet
[main]

# Options only for testnet
[test]
# [rpc]
# Accept command line and JSON-RPC commands.
server=1
# Bind to given address to listen for JSON-RPC connections. This option is ignored unless -rpcallowip
# is also passed. Port is optional and overrides -rpcport. Use [host]:port notation for IPv6.
# This option can be specified multiple times. (default: 127.0.0.1 and ::1 i.e., localhost,
# or if -rpcallowip has been specified, 0.0.0.0 and :: i.e., all addresses)
rpcbind=0.0.0.0
# Listen for JSON-RPC connections on this port
rpcport=18332
# Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4),
# a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This option can be
# specified multiple times.
rpcallowip=0.0.0.0/0
# Number of seconds after which an uncompleted RPC call will time out
rpcservertimeout=3
# Dangerous, only use for testnet
rpcuser=bitcoincorerpc
rpcpassword=bitcoincorerpc

# Options only for regtest
[regtest]
```

#### Run the image
```bash
 docker run -v /root/.bitcoin:/opt/bitcoin --network=host build-test/bitcoind:0.19.0.1
```

#### After the testnet fully in-sync, terminate it, then re-run the docker

You should see it will not the full resync again.

### CODE REVIEW

**Comment by nevrending** [comment]
I'm trying something along
```dockerfile
RUN groupadd -g ${GROUP_ID} bitcoin \
        && useradd -u ${USER_ID} -g bitcoin -s /bin/bash -m -d /home/bitcoin bitcoin
```
And using gosu to step-down from root

File: Dockerfile [hunk 1]
```diff
@@ -19,17 +18,16 @@ RUN echo "\nSubmodule files:"&& \
     make -j8
 RUN ./autogen.sh && \
     ./configure LDFLAGS="-static-libstdc++" --prefix="/app/src/bitcoin/depends/x86_64-pc-linux-gnu" \
-    --enable-glibc-back-compat --enable-hardening --with-zmq --disable-man \
-    --disable-bench --disable-test --without-gui
+    --without-miniupnpc --enable-hardening --with-zmq --disable-man --disable-shared \
```
Comment (dendisuhubdy) [inline Dockerfile]: :+1:

File: Dockerfile [hunk 2]
```diff
@@ -40,21 +38,13 @@ LABEL maintainer.0="Dendi Suhubdy (dendi@bitwyre.com)" \
...
 COPY --from=builder /app/src/bitcoin/depends/x86_64-pc-linux-gnu /usr/local
-COPY docker-entrypoint.sh /docker-entrypoint.sh
-RUN chmod +x /docker-entrypoint.sh
 
-VOLUME ["/home/bitcoin/.bitcoin"]
+VOLUME ["/root/.bitcoin"]
 EXPOSE 8332 8333 18332 18333 18443 18444 28332 28333 28334 28335
 
-ENV BITCOIN_DATA = "/home/bitcoin/.bitcoin"
+ENV BITCOIN_DATA = "/root/.bitcoin"
```
Comment (dendisuhubdy) [inline Dockerfile]: wait why?

Comment (Ujang360) [reply Dockerfile]: Because of this
image

**Comment by Ujang360** [comment]
> I'm trying something along
> 
> ```dockerfile
> RUN groupadd -g ${GROUP_ID} bitcoin \
>         && useradd -u ${USER_ID} -g bitcoin -s /bin/bash -m -d /home/bitcoin bitcoin
> ```
> 
> And using gosu to step-down from root

This means passing thru the host user, fantastic :tada:

**Comment by nevrending** [comment]
> > I'm trying something along
> > ```dockerfile
> > RUN groupadd -g ${GROUP_ID} bitcoin \
> >         && useradd -u ${USER_ID} -g bitcoin -s /bin/bash -m -d /home/bitcoin bitcoin
> > ```
> > 
> > 
> > And using gosu to step-down from root
> 
> This means passing thru the host user, fantastic 🎉
> 
> **Edit:**
> I get that wrong, can you explain more @nevrending ?

`gosu` is basically `su` in go haha
To get out from root and enter shell as uid/gid

**Pull request merged**